projects/step-frontend/src/lib/modules/timeseries/ranger/ts-ranger.component.ts

```diff
@@ -0,0 +1,284 @@
+import {
+  AfterViewInit,
+  Component,
+  ElementRef,
+  EventEmitter,
+  Input,
+  OnInit,
+  Output,
+  ViewChild,
+  ViewEncapsulation,
+} from '@angular/core';
+import { TSRangerSettings } from './ts-ranger-settings';
+import { UplotSyncService } from '../chart/uplot-sync-service';
+import { TSTimeRange } from '../chart/model/ts-time-range';
+import { UPlotUtils } from '../uplot/uPlot.utils';
+import { timeout } from 'rxjs';
+
+declare const uPlot: any;
+
+@Component({
+  selector: 'step-ts-ranger',
+  templateUrl: './ts-ranger.component.html',
+  styleUrls: ['./ts-ranger.component.scss'],
+  encapsulation: ViewEncapsulation.None,
+})
+export class TSRangerComponent implements OnInit, AfterViewInit {
+  @ViewChild('chart') private chartElement: ElementRef;
+
+  @Input('settings') settings: TSRangerSettings;
+  @Input('syncKey') syncKey: string;
+
+  @Output('onRangeChange') onRangeChange = new EventEmitter<TSTimeRange>();
+
+  uplot: any;
+  lastRange: TSTimeRange;
+
+  start: number;
+  end: number;
+
+  ngOnInit(): void {
+    if (this.syncKey) {
+      uPlot.sync(this.syncKey);
+    }
+    this.start = this.settings.xValues[0];
+    this.end = this.settings.xValues[this.settings.xValues.length - 1];
+  }
+
+  ngAfterViewInit(): void {
+    this.createRanger();
+  }
+
+  resetSelect() {
+    // this is a 'hack'. when dblclick is triggered in another synced chart, it will remove the select for the ranger. this function is executed before that.
+    // we have to wait the minimum amount of time so that sync event happens, and the selection is destroyed
+    setTimeout(() => {
+      let left = Math.round(this.uplot.valToPos(this.start, 'x'));
+      let width = Math.round(this.uplot.valToPos(this.end, 'x')) - left;
+      let height = this.uplot.bbox.height / devicePixelRatio;
+      this.uplot.setSelect({ left, width, height }, false);
+      let xData = this.uplot.data[0];
+      let start = xData[0];
+      let end = xData[xData.length - 1];
+      this.onRangeChange.emit({ start, end });
+    }, 50);
+  }
+
+  createRanger() {
+    // @ts-ignore
+    let x0;
+    let lft0: number;
+    let wid0: number;
+    // @ts-ignore
+    const lftWid = { left: null, width: null };
```

Comment:
I picked this one as an example, I dislike abreveated varibles like "lft" or "el" as long as the variables do not become super long oterwise. The "lftWid" variable could be split into "width" and "left" and the nassembled when used like:

```
let select = (newLeft, newWidth) => {
      this.uplot.setSelect({left: newLeft, width: newWidth}, false);
    };
```

Reply:
These functions are not implemented by me, but "part" of the library. My scope was to have it up and running asap, that's why I left it the way it was. Same for ts-ignore. There are @ts-ignore just on the functions written by the library contributors. Of course we can clean all them up.